### ISSUE TITLE
TinyMCE 5 Preview: Create new UI controls by extending existing ones

### ISSUE THREAD

**Issue by ryandemmer** [tinymce/tinymce#4588]
In Tinymce 4.x it is possible to create new UI controls by extending existing controls, eg:

```javascript
var DataList = tinymce.ui.ComboBox.extend({
    init: function() {
         this._super();
    },
    value: function(value) {
        this._super(value);
    }
});
tinymce.ui.Factory.add('DataList', DataList);
```

but this does not appear to be possible in Tinymce 5, as the UI controls are not exposed in the API.

Is there another way this is done in Tinymce 5?

**Comment by TheSpyder** [comment]
You're right, the intention of the new UI is that controls are not exposed in the API directly.

The aim is to provide flexible components that don't need to be extended, making it easier to create dialogs that adjust to the active skin. The reduced API surface will also be easier for us to maintain.

Whether we are able to maintain that separation is something we hope to find out with this Developer Preview :)

Can you tell us more about why you need to extend controls?

**Comment by ryandemmer** [comment]
> You're right, the intention of the new UI is that controls are not exposed in the API directly.
> 
> The aim is to provide flexible components that don't need to be extended, making it easier to create dialogs that adjust to the active skin

That makes the UI less useful, as it restricts developers to the rather limited set of controls that you create an maintain, which are suitable for only simple dialogs. It also prevents developers from creating new controls for the toolbar, or adding features to existing ones.

> Can you tell us more about why you need to extend controls?

Quite simply because the UI, in its present form and in Tinymce 5, does not provide the controls I require to accomplish the tasks that I need them to. Over quite a long period of time, I have created and tested a number of controls to suit my purposes.

Here is an example of some of the controls I have created in use:

![JCE custom controls example](https://cdn.joomlacontenteditor.net/tmp/jce-image-attributes-example.gif)

**Comment by ryandemmer** [comment]
All of these custom controls (and others not pictured) require extension of existing controls. Although Tinymce 5 appears to include versions of some of the controls I have created - ColorPicker, SizeInput, UrlInput - they do not contain the features I require for my use cases.

I'm sure other developers appreciate the ability to extend controls in Tinymce 4 and would also like to see that option in Tinymce 5.

**Comment by ryandemmer** [comment]
Thank you.

I would like to stress that even if the core UI supports many similar controls, it would still be beneficial (or essential in my case) to be able to create custom controls and extend core UI controls. There will always be a case where something extra is required.

Here is the code for my "CustomValue" control demonstrated in the Attributes tab (without the repeatable option) to show the kind of control extension I am referring to:

``` javascript
/**
   * This class creates a CustomValue control.
   *
   * @class tinymce.ui.CustomValue
   * @extends tinymce.ui.FormItem
   */
tinymce.ui.CustomValue = tinymce.ui.FormItem.extend({

    Defaults: {
        spacing: 5
    },

    /**
     * Constructs a instance with the specified settings.
     *
     * @constructor
     * @param {Object} settings Name/value object with settings.
     */
    init: function (settings) {
        var self = this;

        settings.values = settings.values || [];

        settings.size = settings.size || 35;

        self._super(settings);
        self.classes.add('customvalue');
    },

    name: function () {
        var prefix = this.settings.prefix || '';

        return prefix + this._items[0].value();
    },

    /**
     * Getter/setter function for the control value.
     *
     * @method value
     * @param {Object} [value] Value to be set.
     * @return {Object|tinymce.ui.CustomValue} Value or self if it's a set operation.
     */
    value: function (value) {
        if (typeof value !== 'undefined') {

            if (typeof value !== "object") {
                return this;
            }

            this.state.set('value', value);

            if (this.state.get('rendered')) {

                for (var key in value) {
                    this._items[0].value(key);
                    this._items[1].value(value[key]);
                }
            }

            return this;
        }

        // Make sure the real state is in sync
        if (this.state.get('rendered')) {
            value = "";

            var key = this._items[0].value();

            if (key) {
                value = {};
                value[key] = this._items[1].value();
            }

            this.state.set('value', value);
        }

        return this.state.get('value');
    },

    /**
     * Post render method. Called after the control has been rendered to the target.
     *
     * @method postRender
     * @return {tinymce.ui.CustomValue} Current CustomValue instance.
     */
    postRender: function () {
        var self = this,
            id = this._id,
            settings = self.settings;
        self._super();

        var name = {
            type: 'textbox',
            size: settings.size,
            placeholder: 'Name'
        };

        var value = {
            type: 'textbox',
            size: Math.floor(settings.size * 1.25),
            placeholder: 'Value'
        };

        if (settings.values.length) {
            name = {
                type: 'datalist',
                placeholder: 'Name',
                values: settings.values
            }

            value.size = settings.size;
        }

        // create datalist of attribute names
        self.append(name).append(value).reflow();

        return self;
    }
});

tinymce.ui.Factory.add('CustomValue', tinymce.ui.CustomValue);
```

**Comment by TheSpyder** [comment]
Would it be enough if we included a plugin with all of the 4.x UI controls? Dialogs created that way may not match the TinyMCE 5 dialog style (unless you apply that yourself), but that would let you retain the benefits of our improved default dialogs and toolbar while also using custom dialogs as necessary.

**Comment by ryandemmer** [comment]
Does mean that each dialog that contains at least one custom control will have to be built entirely with the old UI? I assume that it would not be possible to mix and match the old and new UI.

Unfortunately this does not solve the original problem, which is that the new UI is not extendable, and therefore any ambitious developer must therefore either accept the limitations of the new UI, or use some other solution to display complex dialogs.

**Comment by cshelleig** [comment]
I see that tinymce.ui is still referenced in the docs, but, is in fact, gone. viz.
https://www.tiny.cloud/docs-beta/api/tinymce.ui/tinymce.ui.factory/
Is there a plan to re-implement tinymce.ui.factory?

**Comment by TheSpyder** [comment]
No, the factory was left in by accident when we deleted the ui folder. The documentation generator is a bit unstable due to the age of the code so the website can be out of date at times.

The release will happen as scheduled without extendable controls, we will evaluate our options to fill this need afterwards.

**Comment by cshelleig** [comment]
I rather assumed so. We create some lighter weight instances of the editor, and/or use the same textarea with a light weight one of our own. Up til now we used methods like
tinymce.ui.Factory.create({
            type: 'buttongroup',
and so on.
Is there a known work around for something like that?
I'm still reading docs and code to see what might work.

**Comment by TheSpyder** [comment]
Are you talking about the toolbar? This discussion was centred around dialogs. I'm wondering whether toolbar customisation could be replaced with tweaks to the inline or quickbars (previously inlite) UIs.

**Comment by cshelleig** [comment]
Ding! my bad. will search around and see what i can see.
thanks, and sorry for polluting the thread.

**Comment by TheSpyder** [comment]
That's ok! We appreciate feedback on this release in all forms ☺️

**Comment by josh18** [comment]
Is there any info around how to customise toolbar buttons? Or is it the same boat as dialogues? TheSpyder mentioned something but I can't really find any other info.

Specifically I want to be able to manage the enable / disable state of built in buttons such as indent without having to rebuild the whole button but I can't see any way to do this.

**Comment by TheSpyder** [comment]
There isn't a supported way, no. I can think of a trick that would make re-registering the built in button easier but it relies on implementation details which may be subject to change in the future.

**Comment by ryandemmer** [comment]
I should comment here by saying that for me this issue has been resolved by the creation of a custom Theme and UI for Tinymce 5, based on the Theme and UI components from Tinymce 4.x

As the Core, Theme and UI are largely separate libraries in Tinymce 5, this was relatively easy to do, but did require a few workarounds.  I had to bypass editor.ui.registry for registering buttons etc. and use a custom implementation located in the Theme. This is to overcome the limitations imposed by the specifications set for toolbar buttons, menu buttons etc. Creating a button now uses

```editor.theme.addButton('name', {});```

instead of 

```editor.ui.registry.addButton('name', {});```

So far it is working very well, and I'm particularly pleased to be using Tinymce 5 as the editor "engine" underneath my custom chassis, rather than having to either stick with Tinymce 4 or a forked of modified version of it.
